cmd/kaf/consume.go

```diff
@@ -4,63 +4,113 @@ import (
 	"bytes"
 	"encoding/binary"
 	"fmt"
+	"math"
 	"os"
+	"sort"
 	"strconv"
+	"strings"
 	"sync"
+	"sync/atomic"
 	"text/tabwriter"
 	"time"
 
 	"github.com/Shopify/sarama"
 	"github.com/birdayz/kaf/avro"
 	"github.com/birdayz/kaf/pkg/proto"
 	"github.com/golang/protobuf/jsonpb"
-	prettyjson "github.com/hokaccha/go-prettyjson"
-	colorable "github.com/mattn/go-colorable"
+	"github.com/hokaccha/go-prettyjson"
+	"github.com/mattn/go-colorable"
 	"github.com/spf13/cobra"
 )
 
 var (
-	offsetFlag  string
-	raw         bool
-	follow      bool
-	schemaCache *avro.SchemaCache
-	keyfmt      *prettyjson.Formatter
+	partitionsWhitelistFlag []int
+	offsetFlags             []string
+	countFlags              []int
+	endFlag                 bool
+	raw                     bool
+	followFlag              bool
+	tailFlag                int
+	schemaCache             *avro.SchemaCache
+	keyfmt                  *prettyjson.Formatter
 
 	protoType string
 )
 
 func init() {
 	rootCmd.AddCommand(consumeCmd)
-	consumeCmd.Flags().StringVar(&offsetFlag, "offset", "oldest", "Offset to start consuming. Possible values: oldest, newest.")
-	consumeCmd.Flags().BoolVar(&raw, "raw", false, "Print raw output of messages, without key or prettified JSON")
-	consumeCmd.Flags().BoolVarP(&follow, "follow", "f", false, "Shorthand to start consuming with offset HEAD-1 on each partition. Overrides --offset flag")
-	consumeCmd.Flags().StringSliceVar(&protoFiles, "proto-include", []string{}, "Path to proto files")
-	consumeCmd.Flags().StringSliceVar(&protoExclude, "proto-exclude", []string{}, "Proto exclusions (path prefixes)")
-	consumeCmd.Flags().StringVar(&protoType, "proto-type", "", "Fully qualified name of the proto message type. Example: com.test.SampleMessage")
+	consumeCmd.Flags().StringSliceVarP(&offsetFlags, "offset", "o", []string{},
+		`Offset to start consuming. Pass a single value to apply to all partitions; pass a comma-delimited list to apply to specific partitions.
+Possible values: oldest, newest, -X, +X, X. (X is a number)
+oldest: Start reading at the oldest offset. This is the default.
+newest: Start reading at the newest offset.
+X: Start reading at offset X.
+-X: Start reading X before newest offset.
++X: Start reading X after oldest offset.`)
+	consumeCmd.Flags().IntSliceVarP(&countFlags, "count", "c", []int{}, "Number of messages to consume before exiting. Specify a single value for global limit, a comma-delimited list to limit specific partitions.")
+	consumeCmd.Flags().IntSliceVarP(&partitionsWhitelistFlag, "partitions", "p", []int{}, "Partitions to read from as a comma-delimited list; if unset, all partitions will be read.")
+	consumeCmd.Flags().BoolVar(&raw, "raw", false, "Print raw output of messages, without key or prettified JSON.")
+	consumeCmd.Flags().BoolVarP(&followFlag, "follow", "f", false, "Shorthand to start consuming with offset HEAD-1 on each partition. Overrides --offset flag.")
```

Comment:
this would change current behavior; currently following is enabled by default. i would rather not change this behavior, so better default to true

Reply:
This flag defaults to `false` in master too, but I found `--follow` to be a little confusing in kaf. I'm used to `--follow` indicating that the command should keep waiting for additional data when it reaches the end of whatever it's reading from (as in `tail` or `kubectl logs`), but kaf is a little different: in kaf the default behavior is to keep waiting for messages, and `--follow` means read from max offset - 1. I added the `--end/-e` flag so that you can tell kaf to *stop* when it reaches the end, but left the semantics of the `--follow` flag the same and kept the default behavior of waiting for messages. (I did break `--follow` initially but I think I fixed it in d6d0c97.)